### PR TITLE
Destination Kafka

### PR DESCRIPTION
**Pull request by prabrisha-rudder**
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/271)

### CODE REVIEW

File: config/sample.env [hunk 1]
```diff
@@ -79,3 +79,6 @@ WAREHOUSE_JOBS_DB_PORT=5432
 WAREHOUSE_JOBS_DB_DB_NAME=jobsdb
 
 WAREHOUSE_URL=http://localhost:8082
+
+SSL_CERTIFICATE_FILE_PATH=<certificate_file_path>
```
Comment (SumanthPuram) [inline config/sample.env:83]: Rename these to KAFKA_SSL_CERTIFICATE_FILE_PATH

File: processor/integrations/integrations.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ func init() {
 
 func loadConfig() {
 	destTransformURL = config.GetEnv("DEST_TRANSFORM_URL", "http://localhost:9090")
-	customDestination = []string{"KINESIS"}
+	customDestination = []string{"KINESIS", "KAFKA"}
```
Comment (SumanthPuram) [inline processor/integrations/integrations.go:26]: Given that we updated transformers, let's clean this function - GetUserIDFromTransformerResponse

File: processor/processor.go [hunk 3]
```diff
@@ -639,7 +641,11 @@ func (proc *HandleT) processJobsForDest(jobList []*jobsdb.JobT, parsedEventList
 							panic(fmt.Errorf("typecast of singularEvent to map[string]interface{} failed"))
 						}
 						shallowEventCopy["message"] = singularEventMap
-						shallowEventCopy["destination"] = reflect.ValueOf(destination).Interface()
+
+						if !misc.ContainsString(customDestinations, destType) {
```
Comment (SumanthPuram) [inline processor/processor.go:645]: Why do we need this?

File: router/router.go [hunk 4]
```diff
@@ -184,7 +203,33 @@ func (rt *HandleT) workerProcess(worker *workerT) {
 
 			deliveryTimeStat.Start()
 			if misc.ContainsString(customDestinations, job.CustomVal) {
-				respStatusCode, respStatus, respBody = customdestinationmanager.Send(job.EventPayload, job.CustomVal)
+				key := paramaters.SourceID + "-" + paramaters.DestinationID
```
Comment (SumanthPuram) [inline router/router.go:206]: We should move this code out of router. May be into custom destination manager.

File: router/router.go [hunk 5]
```diff
@@ -700,6 +741,56 @@ func init() {
...
+						}
+						if destination.Enabled {
+							producer, err := customdestinationmanager.GetProducer(destConfig, rt.destID)
+							producerConfig := ProducerConfig{Config: destConfig, Producer: producer}
+							destinationConfigProducerMap[key] = producerConfig
+							if err == nil {
+								logger.Infof("========== created new producer: %v for destination: %v and source %v", producer, destination.Name, source.Name)
+							} else {
+								producerConfig := ProducerConfig{Config: destConfig, Producer: producer}
+								destinationConfigProducerMap[key] = producerConfig
```
Comment (SumanthPuram) [inline router/router.go:782]: We need to synchronize the access to destinationConfigProducerMap. Here we are writing and reading inside workerProcess.

File: services/streammanager/kafka/kafkamanager.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	err = json.Unmarshal(jsonConfig, &destConfig)
+
+	hostName := destConfig.HostName + ":" + destConfig.Port
+	isSslEnabled := destConfig.SslEnabled
+
+	hosts := []string{hostName}
+
+	config := sarama.NewConfig()
+	config.Producer.Partitioner = sarama.NewReferenceHashPartitioner
+	config.Producer.RequiredAcks = sarama.WaitForAll
```
Comment (SumanthPuram) [inline services/streammanager/kafka/kafkamanager.go:65]: We need to put this in docs (all the values we are using by default). If someone wants to override, we should ask them to contact us, so we can understand their case better.

File: services/streammanager/kafka/kafkamanager.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+func loadConfig() {
+	clientCertFile = config.GetEnv("SSL_CERTIFICATE_FILE_PATH", "")
+	clientKeyFile = config.GetEnv("SSL_KEY_FILE_PATH", "")
+}
+
+func loadCertificate() {
+	if clientCertFile != "" && clientKeyFile != "" {
+		cert, err := tls.LoadX509KeyPair(clientCertFile, clientKeyFile)
+		if err != nil {
+			logger.Errorf("error loading certificate: ", err)
```
Comment (SumanthPuram) [inline services/streammanager/kafka/kafkamanager.go:44]: Let's panic here. There should not be an invalid file.

File: processor/processor.go (hunk 3 above)
Comment (prabrisha-rudder) [reply processor/processor.go:645]: While doing load testing, was getting some error while sending events to transformer - EOF. As while setting up aws MSK, in the doc they asked to store the entire jdk ca certs to store in client truststore. While testing, I put all the ca certs (it was huge) that was there in the truststore to webapp config.  So, while data was being sent to transformer, the size of it was too big. That is why getting the error.
Thought that as we have put text area there in webapp, may be we can get similar error. So, removed it for these destinations as we dont need config in transformer.

File: services/streammanager/kafka/kafkamanager.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	msg := &sarama.ProducerMessage{
+		Topic: topic,
+		Key:   sarama.StringEncoder(key),
+		Value: sarama.StringEncoder(message),
+	}
+
+	return msg
+}
+
+// NewTLSConfig generates a TLS configuration used to authenticate on server withccertificates.
```
Comment (SumanthPuram) [inline services/streammanager/kafka/kafkamanager.go:92]: withccertificates - typo?

File: services/streammanager/kafka/kafkamanager.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+	tlsConfig.BuildNameToCertificate()
+	return &tlsConfig
+}
+
+// CloseProducer closes a given producer
+func CloseProducer(producer interface{}) error {
+	kafkaProducer, ok := producer.(sarama.SyncProducer)
+	if ok {
+		kafkaProducer.Close()
```
Comment (SumanthPuram) [inline services/streammanager/kafka/kafkamanager.go:112]: Does Close have any return value? If yes, let's log the error.

File: services/streammanager/kafka/kafkamanager.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+		return nil
+	}
+	return fmt.Errorf("error while closing producer")
```
Comment (SumanthPuram) [inline services/streammanager/kafka/kafkamanager.go:115]: Change all fmts to logger

File: config/sample.env (hunk 1 above)
Comment (prabrisha-rudder) [reply config/sample.env]: Incorporated.

File: processor/integrations/integrations.go (hunk 2 above)
Comment (prabrisha-rudder) [reply processor/integrations/integrations.go]: Done.

File: router/router.go (hunk 4 above)
Comment (prabrisha-rudder) [reply router/router.go]: Incorporated.

File: router/router.go (hunk 5 above)
Comment (prabrisha-rudder) [reply router/router.go]: Used the same mutex.

File: services/streammanager/kafka/kafkamanager.go (hunk 7 above)
Comment (prabrisha-rudder) [reply services/streammanager/kafka/kafkamanager.go]: Incorporated.

File: services/streammanager/kafka/kafkamanager.go (hunk 8 above)
Comment (prabrisha-rudder) [reply services/streammanager/kafka/kafkamanager.go]: Yes. Changed it.

File: services/streammanager/kafka/kafkamanager.go (hunk 9 above)
Comment (prabrisha-rudder) [reply services/streammanager/kafka/kafkamanager.go]: Incorporated.

File: services/customdestinationmanager/customdestinationmanager.go [hunk 11]
```diff
@@ -2,31 +2,213 @@ package customdestinationmanager
...
+	if producerConfig != (ProducerConfig{}) || producerConfig.Producer != nil {
+		respStatusCode, respStatus, respBody = Send(jsonData, destination, producerConfig.Producer, producerConfig.Config)
+	} else {
+		respStatusCode, respStatus, respBody = 400, "Producer not found in router", "Producer could not be created"
+	}
+
+	return respStatusCode, respStatus, respBody
+}
+
+// CreateOrUpdateProducer creates or updates producer based on destination config
```
Comment (SumanthPuram) [inline services/customdestinationmanager/customdestinationmanager.go:148]: This function should return a producer and error. 
Then update  `destinationConfigProducerMap[key]` in the caller

File: services/customdestinationmanager/customdestinationmanager.go [hunk 12]
```diff
@@ -2,31 +2,213 @@ package customdestinationmanager
...
+			the destinatin, such as Kafka server was not up, so, producer could not be created. But, after sometime it becomes reachable,
+			so, while sending the event if the producer is not creatd alrady, then it tries to create it. */
+			producerCreationLock.Lock()
+			producerConfigFromMap := destinationConfigProducerMap[key]
+			producer = producerConfigFromMap.Producer
+			config = producerConfigFromMap.Config
+			if producer != nil {
+				producerConfig = producerConfigFromMap
+			} else {
+				producer, err := GetProducer(config, destination)
```
Comment (SumanthPuram) [inline services/customdestinationmanager/customdestinationmanager.go:124]: Should we name this NewProducer?

File: services/customdestinationmanager/customdestinationmanager.go [hunk 13]
```diff
@@ -2,31 +2,213 @@ package customdestinationmanager
...
+				}
+			}
+			producerCreationLock.Unlock()
+		}
+	} else {
+		producerCreationLock.RUnlock()
+	}
+
+	if producerConfig != (ProducerConfig{}) || producerConfig.Producer != nil {
+		respStatusCode, respStatus, respBody = Send(jsonData, destination, producerConfig.Producer, producerConfig.Config)
```
Comment (SumanthPuram) [inline services/customdestinationmanager/customdestinationmanager.go:140]: Send can be private now?

File: services/customdestinationmanager/customdestinationmanager.go (hunk 11 above)
Comment (prabrisha-rudder) [reply services/customdestinationmanager/customdestinationmanager.go:148]: It is basically a sub part of backendConfigSubscriber and is called from that function only to update the config-producer map. For creating a new producer, there is a different method, named, newProducer(). Updated the comment. Does it make sense?

File: services/customdestinationmanager/customdestinationmanager.go (hunk 12 above)
Comment (prabrisha-rudder) [reply services/customdestinationmanager/customdestinationmanager.go]: Done.

File: services/customdestinationmanager/customdestinationmanager.go (hunk 13 above)
Comment (prabrisha-rudder) [reply services/customdestinationmanager/customdestinationmanager.go]: Done

**Pull request merged**